### PR TITLE
Feature/multiple undescore test allowed

### PR DESCRIPTION
**Pull request by acando14**
I create a feature to allow the testing methods with snake case notation like this 

- MethodName_StateUnderTest_ExpectedBehavior
- MethodName_ExpectedBehavior_StateUnderTest
- Should_ExpectedBehavior_When_StateUnderTest
- When_StateUnderTest_Expect_ExpectedBehavior
- Given_Preconditions_When_StateUnderTest_Then_ExpectedBehavior

I hope it will be helpfull for someone that use our same testing notation.

Thank you so much, regards

### CODE REVIEW

File: src/test/resources/files/Rule/Controversial/CamelCaseMethodName/testRuleDoesNotApplyForTestMethodWithMultipleUnderscoreWhenAllowed.php [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ * @link http://phpmd.org/
+ */
+
+class testRuleDoesNotApplyForTestMethodWithMultipleUnderscoreWhenAllowed
+{
+    public function test_snake_case_method()
+    {
+
+    }
+}
```
Comment (kylekatarnls) [inline src/test/resources/files/Rule/Controversial/CamelCaseMethodName/testRuleDoesNotApplyForTestMethodWithMultipleUnderscoreWhenAllowed.php]: @acando14 Add a blank line please

File: src/test/resources/files/Rule/Controversial/CamelCaseMethodName/testRuleAppliesToTestMethodWithMultipleUnderscores.php [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+ * @link http://phpmd.org/
+ */
+
+class testRuleAppliesToTestMethodWithMultipleUnderscores
+{
+    public function test_snake_case_method()
+    {
+
+    }
+}
```
Comment (kylekatarnls) [inline src/test/resources/files/Rule/Controversial/CamelCaseMethodName/testRuleAppliesToTestMethodWithMultipleUnderscores.php]: @acando14 Add a blank line please

File: src/main/resources/rulesets/controversial.xml [hunk 3]
```diff
@@ -155,4 +161,4 @@ class ClassName {
             ]]>
         </example>
     </rule>
-</ruleset>
+</ruleset>
```
Comment (kylekatarnls) [inline src/main/resources/rulesets/controversial.xml]: @acando14 Please keep the blank line.

File: src/main/php/PHPMD/Rule/Controversial/CamelCaseMethodName.php [hunk 4]
```diff
@@ -71,6 +71,10 @@ public function apply(AbstractNode $node)
 
     private function isValid($methodName)
     {
+        if ($this->getBooleanProperty('allow-multiple-underscore-test') && strpos($methodName, 'test') === 0) {
```
Comment (ravage84) [inline src/main/php/PHPMD/Rule/Controversial/CamelCaseMethodName.php]: For better readability ``allow-multiple-underscores-test``, underscore with S.
Please, adjust the files below accordingly.

File: src/test/resources/files/Rule/Controversial/CamelCaseMethodName/testRuleAppliesToTestMethodWithMultipleUnderscores.php (hunk 2 above)
Comment (acando14) [reply src/test/resources/files/Rule/Controversial/CamelCaseMethodName/testRuleAppliesToTestMethodWithMultipleUnderscores.php]: thank you! Sorry for my delay

File: src/main/php/PHPMD/Rule/Controversial/CamelCaseMethodName.php (hunk 4 above)
Comment (acando14) [reply src/main/php/PHPMD/Rule/Controversial/CamelCaseMethodName.php]: Done! thank you!

**Comment by ravage84** [comment]
Closing this now as I think this is a case of feature creep.

The following line displays beautifully what's wrong with this feature:

https://github.com/phpmd/phpmd/pull/541/files#diff-9ae0f6886004da4d7d3fcfef794feb7dR21

This is not the camel case notation but clearly the snake case one. The rule is named ``CamelCaseMethodName``, though.

We probably shouldn't have added the previous property ``allow-underscore-test`` in the first place.
Probably an ``ignore-tests`` or something like that would have been more flexible.

Also, PHPMD is not focused on code style but on detecting code mess.
If we clutter PHPMD's rules to allow each and every exception one could have, we only create blind spots.

@acando14 hope you understand and sorry for the  late decision.

**Pull request closed without merging**